module/dkg/broker.go

```diff
@@ -0,0 +1,168 @@
+package dkg
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/crypto"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/model/messages"
+	"github.com/onflow/flow-go/module"
+)
+
+// Broker is an implementation of the DKGBroker interface which is intended to
+// be used in conjuction with the DKG MessagingEngine for private messages, and
+// with the DKG smart-contract for broadcast messages.
+type Broker struct {
+	log               zerolog.Logger
+	dkgInstanceID     string                   // unique identifier of the current dkg run (prevent replay attacks)
+	committee         flow.IdentifierList      // IDs of DKG members
+	myIndex           int                      // index of this instance in the committee
+	dkgContractClient module.DKGContractClient // client to communicate with the DKG smart contract
+	tunnel            *BrokerTunnel            // channels through which the broker communicates with the network engine
+	msgCh             chan messages.DKGMessage // channel to forward incoming messages to consumers
+	messageOffset     uint                     // offset for next broadcast messages to fetch
+}
+
+// NewBroker instantiates a new epoch-specific broker capable of communicating
+// with other nodes via a network engine and dkg smart-contract.
+func NewBroker(
+	log zerolog.Logger,
+	dkgInstanceID string,
+	committee flow.IdentifierList,
+	myIndex int,
+	dkgContractClient module.DKGContractClient,
+	tunnel *BrokerTunnel) *Broker {
+
+	b := &Broker{
+		log:               log.With().Str("component", "broker").Str("dkg_instance_id", dkgInstanceID).Logger(),
+		dkgInstanceID:     dkgInstanceID,
+		committee:         committee,
+		myIndex:           myIndex,
+		dkgContractClient: dkgContractClient,
+		tunnel:            tunnel,
+		msgCh:             make(chan messages.DKGMessage),
+	}
+
+	go b.listen()
+
+	return b
+}
+
+/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Implement DKGBroker
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+
+// PrivateSend sends a DKGMessage to a destination over a private channel. It
+// appends the current DKG instance ID to the message.
+func (b *Broker) PrivateSend(dest int, data []byte) {
+	if dest >= len(b.committee) || dest < 0 {
+		b.log.Error().Msgf("destination id out of range: %d", dest)
+		return
+	}
+	dkgMessageOut := messages.DKGMessageOut{
+		DKGMessage: messages.NewDKGMessage(b.myIndex, data, b.dkgInstanceID),
+		DestID:     b.committee[dest],
+	}
+	b.tunnel.SendOut(dkgMessageOut)
+}
+
+// Broadcast broadcasts a message to all participants.
+func (b *Broker) Broadcast(data []byte) {
+	dkgMessage := messages.NewDKGMessage(
+		b.myIndex,
+		data,
+		b.dkgInstanceID,
+	)
+	err := b.dkgContractClient.Broadcast(dkgMessage)
+	if err != nil {
+		if err != nil {
+			b.log.Error().
```

Comment:
Broadcast implements the DKGProcessor interface. It is called directly by @tarakby 's DKG library, and it doesnt have an error return value. 
We could decide to add an error value to the method, and have Tarak's library handle it.. what do you think @tarakby ?
Otherwise I would be happy to implement a retry logic in this function.. any ideas on number and frequency of retries?

Reply:
If the the error from `Broadcast` is returned to the crypto package, the dkg instance has no way to recover and would just shut the instance down. Joint-Feldman protocol doesn't treat the case where a node is not able to communicate, that node would just be considered malicious. I also think implementing the retry on the crypto level doesn't make quite sense as the library is agnostic of the communication channels. 

Shouldn't a retry mechanism be implemented in the contract client itself? I guess the `broker` is also agnostic of the communication channel used. If the broker gets back an error (unable to broadcast), we should just log the issue and carry on. Nodes that couldn't broadcast will be considered malicious but the entire protocol can still remain alive.

Reply:
Right, I agree that the crypto package shouldn't do the retrying.

@tarakby if a broadcast from node `n` fails, everyone else will consider `n` malicious. But will `n` still be able to generate its private key and successfully complete the protocol? Just trying to understand the impact of a broadcast failure and how much we should invest in trying to avoid them right now.

Reply:
There are different scenarios depending on what message could not be broadcasted.

The node itself will complete the protocol and will end up having its own wrong private key. It will be used during the epoch to sign random beacon votes but it will fait all verifications by other honest nodes. 
The system remains alive as long as there are not many malicious nodes. That's why we should try as much as we can to save nodes from not being able to broadcast.